File: week6/14891/14891.cpp
#include <iostream>
#include <string>
#include <vector>
#include <math.h>

using namespace std;

class Gear {
private:
    string gear;
    int topidx = 0;
public:
    void set_gear(string);
    int get_top();
    int get_left();
    int get_right();
    void rotate_clockwise();
    void rotate_counterclockwise();
};

void Gear::set_gear(string gear) {
    this->gear = gear;
}
int Gear::get_top() {
    return gear[topidx];
}
int Gear::get_left() {
    return gear[(topidx + 6) % 8];
}
int Gear::get_right() {
    return gear[(topidx + 2) % 8];
}
void Gear::rotate_clockwise() {
    topidx = (topidx + 7) % 8;
}
void Gear::rotate_counterclockwise() {
    topidx = (topidx + 1) % 8;
}

Gear gears[4];
int K;

void set_input();
void rotate_all(int[]);
int cal_total_score();

int main() {
    ios_base::sync_with_stdio(0);
    set_input();

    int n, dir;
    for (int i=0; i<K; i++) {
        cin >> n >> dir;
        n--;
        int rotate[4] = {0, 0, 0, 0};
        rotate[n] = dir;
        for (int left=n-1; left>=0; left--) {
            if (gears[left].get_right() == gears[left+1].get_left())
                break;
            rotate[left] = rotate[left+1] * -1;
        }
        for (int right=n+1; right<4; right++) {
            if (gears[right].get_left() == gears[right-1].get_right())
                break;
            rotate[right] = rotate[right-1] * -1;
        }
        rotate_all(rotate);
    }
    int ans = cal_total_score();
    cout << ans << endl;
}

void set_input() {
    for (int i=0; i<4; i++) {
        string gear;
        cin >> gear;
        gears[i].set_gear(gear);
    }
    cin >> K;
}

void rotate_all(int rotate[]) {
    for (int i=0; i<4; i++) {
        if (rotate[i] == 1)
            gears[i].rotate_clockwise();
        else if (rotate[i] == -1)
            gears[i].rotate_counterclockwise();
    }
}

int cal_total_score() {
    int total_score = 0;
    for (int i=0; i<4; i++) {
        if (gears[i].get_top() == '1')
            total_score += pow(2, i);
    }
    return total_score;
}

File: friends/week6/14891/14891.py
class Gear:
    def __init__(self, gear):
        self.gear = gear
        self.topidx = 0
        
    def rotate_clockwise(self):
        self.topidx = (self.topidx + 7) % 8

    def rotate_counterclockwise(self):
        self.topidx = (self.topidx + 1) % 8

    def get_top(self):
        return int(self.gear[self.topidx])

    def get_left(self):
        return int(self.gear[(self.topidx + 6) % 8])
        
    def get_right(self):
        return int(self.gear[(self.topidx + 2) % 8])

class P_14891:
    def __init__(self):
        self.gears = []
        for _ in range(4):
            self.gears.append(Gear(input()))
        self.K = int(input())
        self.opers = []
        for _ in range(self.K):
            n, d = map(int, input().split())
            self.opers.append((n-1, d))

    def run(self):
        for n, d in self.opers:
            rotate = [0, 0, 0, 0]
            rotate[n] = d
            for left in reversed(range(n)):
                if self.gears[left].get_right() == self.gears[left+1].get_left():
                    break
                rotate[left] = rotate[left + 1] * -1
            for right in range(n+1, 4):
                if self.gears[right].get_left() == self.gears[right-1].get_right():
                    break
                rotate[right] = rotate[right - 1] * -1
            self.rotate_all(rotate)
        ans = self.cal_score()
        print(ans)

    def rotate_all(self, rotate):
        for idx, dir in enumerate(rotate):
            if dir == 1:
                self.gears[idx].rotate_clockwise()
            elif dir == -1:
                self.gears[idx].rotate_counterclockwise()
    
    def cal_score(self):
        total_score = 0
        for i in range(4):
            total_score += pow(2, i) * self.gears[i].get_top()
        return total_score


if __name__ == "__main__":
    P = P_14891()
    P.run()